plugins/nodes/src/js/components/NodesTable.tsx

```diff
@@ -0,0 +1,201 @@
+  { sortKey: "mem", component: NodesTableMemColumn, displayName: "Mem" },
+  { component: NodesTableDiskBarColumn, displayName: "DiskBar" },
+  { sortKey: "disk", component: NodesTableDiskColumn, displayName: "Disk" },
+  { component: NodesTableGPUBarColumn, displayName: "GPUBar" },
+  { sortKey: "gpu", component: NodesTableGPUColumn, displayName: "GPU" }
+];
+
+interface NodesTableProps {
+  hosts: NodesList;
+  nodeHealthResponse: boolean;
+  masterRegion: string;
+}
+
+interface NodesTableState {
+  data: Node[];
+  sortDirection: SortDirection;
+  sortColumn: string;
+}
+
+type SortFunction<T> = (data: T[], sortDirection: SortDirection) => T[];
+
+export default class NodesTable extends React.Component<
+  NodesTableProps,
+  NodesTableState
+> {
+  constructor() {
+    super();
+
+    this.state = {
+      data: [],
+      sortColumn: "hostname",
+      sortDirection: "ASC"
+    };
+
+    this.handleSortClick = this.handleSortClick.bind(this);
+  }
+
+  retrieveSortFunction(sortColumn: string): SortFunction<Node> {
+    const selectedColumn = columns.find(
+      column => column.sortKey === sortColumn
+    );
+
+    if (selectedColumn && selectedColumn.component.sorter) {
+      return selectedColumn.component.sorter;
+    }
+    return data => data;
+  }
+
+  updateData(
+    data: Node[],
+    sortColumn: string,
+    sortDirection: SortDirection,
+    currentSortDirection?: SortDirection,
+    currentSortColumn?: string
+  ): NodesTableState {
+    const copiedData = data.slice();
+
+    if (
+      sortDirection === currentSortDirection &&
+      sortColumn === currentSortColumn
+    ) {
+      return { data: copiedData, sortDirection, sortColumn };
+    }
+
+    if (
+      sortDirection !== currentSortDirection &&
+      sortColumn === currentSortColumn
+    ) {
+      return { data: copiedData.reverse(), sortDirection, sortColumn };
+    }
+
+    const sortFunction = this.retrieveSortFunction(sortColumn);
+
+    return {
+      data: sortFunction(copiedData, sortDirection),
+      sortDirection,
+      sortColumn
+    };
+  }
+
+  handleSortClick(columnName: string): void {
+    const toggledDirection =
+      this.state.sortDirection === "ASC" ? "DESC" : "ASC";
+
+    this.setState(
+      this.updateData(
+        this.state.data,
+        columnName,
+        toggledDirection,
+        this.state.sortDirection,
+        this.state.sortColumn
+      )
+    );
+  }
+
+  componentWillReceiveProps(nextProps: NodesTableProps): void {
+    const items = nextProps.hosts.getItems();
+    items.forEach(item =>
+      item.setEqualToMasterRegion(
+        item.getRegionName() === nextProps.masterRegion
+      )
+    );
+
+    this.setState(
+      this.updateData(items, this.state.sortColumn, this.state.sortDirection)
+    );
+  }
+
+  render() {
+    const { data, sortColumn, sortDirection } = this.state;
+
+    return (
+      <Table data={data.slice()}>
+        {columns.map(
+          column =>
+            column.sortKey ? (
```

Comment:
Hm, I think here's something wrong 🤔

Thinking of this table as an example for others, we cant assume that `sortKey` and "visible Header" are tied together. It might be the case in this particular example (and it isnt anymore, after we integrated the `*bar` columns) - but there might be different tables somewhere - also there's a bug in this refactoring :/ 

That said, I think this whole code block should look like this (sorry for just presenting a solution here, had to think about it for a while…):
```
{columns.map(column => {
  return (
    <Column
      key={column.id} // this should be mandatory and _used as sortKey_ if sortable
      header={getColumnHeader(column, this.handleSortClick, sortColumn === column.id ? sortDirection : null)} // this will have a bit ore logic now…
      cellRenderer={column.component.renderer}
      width={column.component.sizer}
    />
  );
})}
```
`getColumnHeader` will look something like this (this should/could be some kind of a UI Kit helper…)
```
function getColumnHeader(
  column: NodesTableColumn,
  handleSortClick: (columnName: string) => void,
  sortDirection: SortDirection | null
): React.ReactNode {
  if (column.displayName !== undefined && column.sortable === true)
    return (
      <SortableColumnHeader
        columnContent={column.displayName}
        sortHandler={handleSortClick.bind(null, column.id)}
        sortDirection={sortDirection}
      />
    );

  return <HeaderCell>{column.displayName}</HeaderCell>;
}
``

Reply:
actually, there is an even better version of `getColumnHEader` where the `< SortableColumnHeader />` only returns the _contents_ of `<HeaderCell />`. I think that would be even better 👍